File: backend/core/domain/minimax/minimax.py
from core.domain.engine.Position import Position
from core.domain.engine.enums import PieceColor
from core.domain.evaluation.evaluation import evaluate_position


def minimax(position: Position, alpha, beta, depth=5, branches_to_check=10, is_root=True):
    global visited_nodes
    visited_nodes += 1
    if position.is_game_over():

        outcome = position.determine_outcome()
        if outcome == "win":
            return (float('inf'), None) if position.side_to_move == PieceColor.WHITE else (-float('inf'), None)

        elif outcome == "loss":
            return (-float('inf'), None) if position.side_to_move == PieceColor.WHITE else (float('inf'), None)

        else:
            return 0, None

    if depth == 0:
        return evaluate_position(position), None

    best_move = None

    if position.side_to_move == PieceColor.WHITE:
        max_eval = -float('inf')

        # TODO: when MVVLVA will be ready, then no such thing needed --> [:branches_to_check]
        all_moves = position.get_moves_sorted_by_priority()[:branches_to_check]
        # take :branches_to_check with priority 0 and all the rest with bigger priority

        for move in all_moves:
            position_copy = position.clone()
            position_copy.apply_move(move)

            eval, _ = minimax(position_copy, alpha, beta, depth - 1, branches_to_check-1, is_root=False)
            if depth == 4:
                print(eval, move)
            if eval > max_eval:
                max_eval = eval
                best_move = move if is_root else None  # Store the move only on the root call

            alpha = max(alpha, eval)
            if beta <= alpha:
                break
        return max_eval, best_move
    else:  # position.side_to_move == PieceColor.BLACK
        min_eval = float('inf')

        all_moves = position.get_moves_sorted_by_priority()[:branches_to_check]

        for move in all_moves:
            position_copy = position.clone()
            position_copy.apply_move(move)

            eval, _ = minimax(position_copy, alpha, beta, depth - 1, branches_to_check-1, is_root=False)

            if eval < min_eval:
                min_eval = eval
                best_move = move if is_root else None  # Store the move only on the root call

            beta = min(beta, eval)
            if beta <= alpha:
                break
        return min_eval, best_move

File: ChessEngineBackend/ChessEngine.Application/Minimax/Minimax.cs
using System;
using System.Collections.Generic;
using ChessEngine.Enums;
using ChessEngine.Domain.Models;
using ChessEngine.Evaluation;
using System.Diagnostics;
using ChessEngine.FEN;
using ChessEngine.Domain.Interfaces;


namespace ChessEngine.Application.Minimax
{
    public class MinimaxMoveFinder : IMoveFinder
    {
        public (int evaluation, MoveDetail bestMove) GetBestMove(Position position, int depth)
        {
            int alpha, beta;
            if (position.SideToMove == PieceColor.WHITE)
            {
                alpha = int.MaxValue;
                beta = int.MinValue;
            }
            else
            {
                alpha = int.MinValue;
                beta = int.MaxValue;
            }
            Console.WriteLine(depth + " " + alpha + " " + beta);
            return RunMinimax(position, alpha: alpha, beta: beta, depth: depth, isRoot: true);
        }

        private static (int, MoveDetail) RunMinimax(Position position, int alpha, int beta, int depth = 5, bool isRoot = true)  //  int branchesToCheck = 10
        {
            if (position.IsGameOver())
            {
                PositionOutcome outcome = position.DetermineOutcome();
                if (outcome == PositionOutcome.WIN)
                {
                    if (position.SideToMove == PieceColor.WHITE)
                    {
                        return (int.MaxValue, null);
                    }
                    else
                    {
                        return (int.MinValue, null);
                    }
                }
                else if (outcome == PositionOutcome.LOSS)
                {
                    if (position.SideToMove == PieceColor.WHITE)
                    {
                        return (int.MinValue, null);
                    }
                    else
                    {
                        return (int.MaxValue, null);
                    }
                }
                else
                {
                    return (0, null); // Draw
                }
            }

            if (depth == 0)
            {
                int evaluation = Evaluator.EvaluatePosition(position);
                return (evaluation, null);
            }

            MoveDetail bestMove = null;

            if (position.SideToMove == PieceColor.WHITE)
            {
                int maxEval = int.MinValue;

                // TODO: test here allMovesSorted
                List<MoveDetail> allMoves = position.GetAllMoves();
                allMoves.Sort((a, b) =>
                {
                    Position posA = position.Clone();
                    Position posB = position.Clone();
                    posA.ApplyMove(a);
                    posB.ApplyMove(b);
                    return Evaluator.EvaluatePosition(posA).CompareTo(Evaluator.EvaluatePosition(posB));
                });
                // allMoves = allMoves.GetRange(0, Math.Min(branchesToCheck, allMoves.Count));

                foreach (var move in allMoves)
                {
                    Position positionCopy = position.Clone();
                    positionCopy.ApplyMove(move);

                    (int eval, _) = RunMinimax(positionCopy, alpha, beta, depth - 1, isRoot: false); //branchesToCheck - 1,
                    if (eval > maxEval)
                    {
                        maxEval = eval;
                        if (isRoot)
                        {
                            bestMove = move;
                        }
                    }

                    alpha = Math.Max(alpha, eval);
                    if (beta <= alpha)
                    {
                        break;
                    }
                }
                return (maxEval, bestMove);
            }
            else // Black's turn
            {
                int minEval = int.MaxValue;

                // TODO: the same 
                List<MoveDetail> allMoves = position.GetAllMoves();
                allMoves.Sort((a, b) =>
                {
                    Position posA = position.Clone();
                    Position posB = position.Clone();
                    posA.ApplyMove(a);
                    posB.ApplyMove(b);
                    return Evaluator.EvaluatePosition(posA).CompareTo(Evaluator.EvaluatePosition(posB));
                });
                // allMoves = allMoves.GetRange(0, Math.Min(branchesToCheck, allMoves.Count));

                foreach (var move in allMoves)
                {
                    Position positionCopy = position.Clone();
                    positionCopy.ApplyMove(move);

                    (int eval, _) = RunMinimax(positionCopy, alpha, beta, depth - 1, isRoot: false); //  branchesToCheck - 1,
                    if (eval < minEval)
                    {
                        minEval = eval;
                        if (isRoot)
                        {
                            bestMove = move;
                        }
                    }

                    beta = Math.Min(beta, eval);
                    if (beta <= alpha)
                    {
                        break;
                    }
                }
                return (minEval, bestMove);
            }
        }

        //public static int CountMoves(Position position, int depth = 5, bool isRoot = false)
        //{
        //    if (depth == 0)
        //    {
        //        return 1;
        //    }

        //    int count = 0;
        //    List<MoveDetail> moves = position.GetAllSeparateMoves();
        //    foreach (var move in moves)
        //    {
        //        Position copyPosition = position.Clone();
        //        if (isRoot)
        //        {
        //            Console.Write($"{move.Piece} {move.Color} {ConvertToAlgebraic(move.Square, move.Move)}   ");
        //        }
        //        copyPosition.ApplyMove(move);
        //        int copyPositionCount = CountMoves(copyPosition, depth - 1);
        //        count += copyPositionCount;
        //        if (isRoot)
        //        {
        //            Console.WriteLine(copyPositionCount);
        //        }
        //    }

        //    return count;
        //}
    }
}
